core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3867,9 +3719,159 @@ object Stream extends StreamLowPriority {
       * returns `None` instead of `Some(nextStream)`.
       */
     def repeatPull[O2](
-        using: Stream.ToPull[F, O] => Pull[F, O2, Option[Stream[F, O]]]
+        f: Stream.ToPull[F, O] => Pull[F, O2, Option[Stream[F, O]]]
     ): Stream[F, O2] =
-      Pull.loop(using.andThen(_.map(_.map(_.pull))))(pull).void.stream
+      Pull.loop(f.andThen(_.map(_.map(_.pull))))(pull).void.stream
+  }
+
+  implicit final class NestedStreamOps[F[_], O](private val outer: Stream[F, Stream[F, O]])
+      extends AnyVal {
+
+    /**
+      * Nondeterministically merges a stream of streams (`outer`) in to a single stream,
+      * opening at most `maxOpen` streams at any point in time.
+      *
+      * The outer stream is evaluated and each resulting inner stream is run concurrently,
+      * up to `maxOpen` stream. Once this limit is reached, evaluation of the outer stream
+      * is paused until one or more inner streams finish evaluating.
+      *
+      * When the outer stream stops gracefully, all inner streams continue to run,
+      * resulting in a stream that will stop when all inner streams finish
+      * their evaluation.
+      *
+      * When the outer stream fails, evaluation of all inner streams is interrupted
+      * and the resulting stream will fail with same failure.
+      *
+      * When any of the inner streams fail, then the outer stream and all other inner
+      * streams are interrupted, resulting in stream that fails with the error of the
+      * stream that caused initial failure.
+      *
+      * Finalizers on each inner stream are run at the end of the inner stream,
+      * concurrently with other stream computations.
+      *
+      * Finalizers on the outer stream are run after all inner streams have been pulled
+      * from the outer stream but not before all inner streams terminate -- hence finalizers on the outer stream will run
+      * AFTER the LAST finalizer on the very last inner stream.
+      *
+      * Finalizers on the returned stream are run after the outer stream has finished
+      * and all open inner streams have finished.
+      *
+      * @param maxOpen    Maximum number of open inner streams at any time. Must be > 0.
+      */
+    def parJoin(
```

Comment:
This had to move to an invariant `NestedStreamOps` due to Dotty complaining about the type signature of old `parJoin`:

```scala
[error] -- [E043] Type Error: /Users/mpilquist/Development/oss/fs2/core/shared/src/main/scala/fs2/Stream.scala:1759:44
[error] 1759 |  )(implicit ev: O <:< Stream[F2, O2], ev2: F[_] <:< F2[_], F2: Concurrent[F2]): Stream[F2, O2] =  ???
[error]      |                                            ^^^^
[error]      |unreducible application of higher-kinded type F to wildcard arguments
[error] -- [E043] Type Error: /Users/mpilquist/Development/oss/fs2/core/shared/src/main/scala/fs2/Stream.scala:1759:53
[error] 1759 |  )(implicit ev: O <:< Stream[F2, O2], ev2: F[_] <:< F2[_], F2: Concurrent[F2]): Stream[F2, O2] =  ???
[error]      |                                                     ^^^^^
[error]      |unreducible application of higher-kinded type F2 to wildcard arguments
```